### PR TITLE
Fastboot support

### PR DESCRIPTION
**Pull request by mike-north**

### CODE REVIEW

**Comment by buritica** [comment]
Thanks! I haven't had a chance to check out fastboot so I don't know the issue, but is there a way of adding support without a conditional in each function?

**Comment by mike-north** [comment]
The issue is referencing the `Stripe` global. If you had it encapsulated in a module. Essentially we don't have the external script downloaded on the server side, so there is no `Stripe`. Thus, we need to guard everything on the NPM side with `process.env. EMBER_CLI_FASTBOOT`, and everything in the browser  with the `FastBoot` global (the only global that exists in fastboot, including window)

Perhaps a better approach long-term would be to consume `Stripe` as an es6 module, and then the guards could be taken care of within that module.

**Comment by buritica** [comment]
What if the service wasn't initialized if the Stripe global is not available?

Stripe doesn't distribute their library as a module, and I assume they never will for security reasons, so loading it other than globally is not an option.

**Comment by mike-north** [comment]
@buritica consider thinking about it a little differently. The problem is not that `window.Stripe` exists, but rather that the addon consumes it as a global everywhere.

**Comment by buritica** [comment]
ok, i'll have to think about it, thanks for the initial PR. I'll get to it hopefully over the weekend

**Comment by mike-north** [comment]
Something like this might work

**app/stripe.js**
```js
import Ember from 'ember';

const f =  function() {};
const fakeStripe = {
  card: {
    cardType: f,
    validateCardNumber: f,
    validateCVC: f,
    validateExpiry: f,
    createToken() {
      Ember.Logger.info('def');

    }
  },
  bankAccount: {
    createToken() {
      Ember.Logger.info('abc');
    }
  }
};

export default (false && typeof FastBoot === 'undefined') ? window.Stripe : fakeStripe;
```

**addon/services/stripe.js**

```js
import Stripe from 'ember-stripe-services/stripe';
..

```

**Comment by buritica** [comment]
I like that alternative, just got back from some travel. I'll find some time to explore, or if you'd like to update your PR with that including some tests I'll get it out as soon as I can.

I think this could also work for offline testing aswell

**Comment by mike-north** [comment]
@buritica we don't have a great story for testing w/ fastboot yet (you can track current progress here https://github.com/ember-fastboot/ember-cli-fastboot/issues/111 )

**Comment by jakeleboeuf** [comment]
Any chance we can get this merged in, or some other Fastboot support if you don't like @mike-north's approach?

**Comment by buritica** [comment]
@jakeleboeuf I think we never got to include the updated suggestion or tests? I'd be more than happy to add support for this, since I think it could also help for offline testing. right now i'm exremely buried in work, but will code-review and merge this if you can help with updates

**Comment by JoshSmith** [comment]
@mike-north did you want to fold your suggestions into this PR and we can take a look? @jakeleboeuf per @buritica's suggestion you could also feel free to attempt something.

Either myself or one of the other Ember folks in @code-corps/volunteers can try to look into this, as well.

Fastboot is fast becoming a priority for us internally and I'm not well-versed on it yet personally, so there may be some delay until I can get myself up to speed.

**Comment by JoshSmith** [comment]
@mike-north just wanted to check in on this.

**Comment by buritica** [comment]
ping :D

**Comment by mike-north** [comment]
My OSS backlog is rather huge right now, so I'm unable to move this forward for the time being

**Comment by JoshSmith** [comment]
Thanks @mike-north I figured as much. When you do have time – and/or I have time to finally familiarize myself with Fastboot – would be great to just sync up and be sure that if I make changes, that I've implemented them in line with your vision.

Probably at least a few weeks away for me, so if anyone else wants to attempt this (@buritica?) let me know!

File: index.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ module.exports = {
   name: 'ember-stripe-service',
   contentFor: function(type) {
     // we use body since Stripe must be available before
-    if (type === 'body') {
+    if (type === 'body' && process.env.EMBER_CLI_FASTBOOT !== 'true') {
```
Comment (mixonic) [inline index.js]: Why this line?

Comment (mike-north) [inline index.js]: been a while, but I believe the stripe javascript API assumes the existence of globals we don't have in Fastboot

**Comment by buritica** [comment]
Hey everyone, I've asked @sescobb27 and @nicollejimenez to see if they can help us move this forward.  should they pick it up on this PR, or should we try a different approach?

**Comment by JoshSmith** [comment]
It's a pretty small PR @buritica so I'm fine either way. Whatever's easiest for them IMO, and I doubt Mike will mind much if they prefer to start fresh.

**Comment by buritica** [comment]
closed by #55

**Pull request closed without merging**